### ISSUE TITLE
Create a globalwhitelist based on the user object, including claims

### ISSUE THREAD

**Issue by artagel** [vouch/vouch-proxy#96]
If #95 is completed, we can change the "whiteList" config item to a globalWhiteList item, which would be able to validate a user based on anything in the user object.  
That could be username, or any custom claim stored by the user.
This would allow a user to globally allow access if any user is in a group, for example.

**Comment by bnfinet** [comment]
We've been having a slow but extended conversation around `vouch.domains` and `vouch.whitelist` and `vouch.allowAllUsers` and what might be a clearer interface.  I'd like to clean that up to make it more understandable and approachable, probably in a `v1.0.0-pre` branch.

@aaronpk and @halkeye and @simongottschlag have chimed in on such in various issues and on the irc channel (`#vouch` on freenode).

I wouldn't want to break current configs before that, so I'm hesitant to replace `whiteList` at this time.

My other goal these past six months or so has been to see how small I can get the default config.  I'd prefer to keep the configuration surface small whenever possible.  Adding knobs means we'll get more folks showing up in irc and github with support requests.  Of course that's (hopefully) inevitable as the project gets used more but this stuff is already eating into my time so I have an eye on simplifying things wherever possible and making the software self supporting in the config, documentation and in the log messages it provides.

a small example, how do we make this config...
```
   globalClaimWhiteList:
       - claim: groups
         allowed:
           - WebApp Users
           - WebApp Admins
```

as clear as this:
```
    whiteList:
        - bob@yourdomain.com
        - alice@yourdomain.com
        - joe@yourdomain.com
```

And I'm uncertain of where the line should be with Vouch Proxy with regard to authorization.  Should it just leave it to the IdP to authz the user?  If we start going down the path of putting more authority into Vouch Proxy are we on the hook for implementing the OAuth2 spec?  Is Vouch Proxy becoming an IdP?  I know that's a bit philosophical but I've been wondering myself.

**Comment by artagel** [comment]
So, it seems to me that you already jumped the line over to authorization by adding the whitelist to begin with.  That is essentially an authorization function.
In my eyes, vouch provides a beyond corp style system, or at least it provides it partially right now.  It has the proxy piece, it has the authentication piece, but it is missing the authorization piece (for the most part).

I've tried several different options for beyondcorp that are open source, and all of them miss the mark. Mostly because everyone is focused on only the big oauth providers, and most aren't doing generic OIDC like vouch does.  I, and likely many others, need generic OIDC for self-hosted federated auth such as keycloak, adfs, etc.  This is especially useful in environments that are offline, where the public providers simply aren't available.  

As per the comment on the simplicity of configs, I agree and would also like to see a simple config.  YAML can be a bit harder to read IMO, so the same global whitelist in JSON looks like this:
```
{
  "globalClaimWhiteList": [
    {
      "claim": "groups",
      "allowed": [
        "WebApp Users",
        "WebApp Admins"
      ]
    }
  ]
}
```

Perhaps something like this is easier, as long as the config parser can handle it:
`globalClaimWhiteList:
       - groups:
           - WebApp Users
           - WebApp Admins`
{
```
  "globalClaimWhiteList": [
    {
      "groups": [
        "WebApp Users",
        "WebApp Admins"
      ]
    }
  ]
}
```

I've jumped onto the irc so hopefully we can have a discussion on there in more detail.
